.test-infra/mock-apis/src/main/go/cmd/service/echo/main.go

```diff
@@ -0,0 +1,148 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// echo is an executable that runs the echov1.EchoService.
+package main
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+
+	gcplogging "cloud.google.com/go/logging"
+	"github.com/apache/beam/test-infra/mock-apis/src/main/go/internal/cache"
+	"github.com/apache/beam/test-infra/mock-apis/src/main/go/internal/environment"
+	"github.com/apache/beam/test-infra/mock-apis/src/main/go/internal/logging"
+	"github.com/apache/beam/test-infra/mock-apis/src/main/go/internal/service/echo"
+	"github.com/redis/go-redis/v9"
+	"google.golang.org/grpc"
+)
+
+var (
+	env = []environment.Variable{
+		environment.CacheHost,
+		environment.GrpcPort,
+		environment.HttpPort,
+	}
+
+	logger   *slog.Logger
+	logAttrs []slog.Attr
+	opts     = &logging.Options{
+		Name: "echo",
+	}
+)
+
+func init() {
+	for _, v := range env {
+		logAttrs = append(logAttrs, slog.Attr{
+			Key:   v.Key(),
+			Value: slog.StringValue(v.Value()),
+		})
+	}
+}
+
+func main() {
+	ctx := context.Background()
+
+	if !environment.ProjectId.Missing() {
+		client, err := gcplogging.NewClient(ctx, environment.ProjectId.Value())
+		if err != nil {
+			slog.LogAttrs(ctx, slog.LevelError, err.Error(), logAttrs...)
+			os.Exit(1)
+		}
+
+		opts.Client = client
+	}
+
+	logger = logging.New(opts)
+
+	if err := run(ctx); err != nil {
+		logger.LogAttrs(ctx, slog.LevelError, err.Error(), logAttrs...)
+		os.Exit(1)
+	}
+}
+
+func run(ctx context.Context) error {
+	ctx, cancel := signal.NotifyContext(ctx, os.Interrupt)
+	defer cancel()
+
+	if err := environment.Missing(env...); err != nil {
+		return err
+	}
+
+	grpcPort, err := environment.GrpcPort.Int()
+	if err != nil {
+		return err
+	}
+	grpcAddress := fmt.Sprintf(":%v", grpcPort)
+
+	httpPort, err := environment.HttpPort.Int()
+	if err != nil {
+		return err
+	}
+	httpAddress := fmt.Sprintf(":%v", httpPort)
+
+	s := grpc.NewServer()
+	defer s.GracefulStop()
+
+	r := redis.NewClient(&redis.Options{
+		Addr: environment.CacheHost.Value(),
+	})
+
+	echoOpts := &echo.Options{
+		Decrementer:  (*cache.RedisCache)(r),
+		LoggingAttrs: logAttrs,
+		Logger:       logger,
```

Comment:
If the intent here is to always print out these attrs with every request, just build them into the logger instead of forcing per-callsite behavior on yourself.

https://pkg.go.dev/log/slog#Logger.With

If the attributes are only going to be logged sometimes though, document the field with the conditions they're logged, so it's clear what that field is for.

Reply:
Ah this is a good tip thank you. I won't change anything with slog related as you requested not up undo my work. I'll keep this comment open as a future reference.